File: webapps/frontend/app/viewmodels/containers/diskpartition.js
// Copyright (C) 2016 iNuron NV
//
// This file is part of Open vStorage Open Source Edition (OSE),
// as available from
//
//      http://www.openvstorage.org and
//      http://www.openvstorage.com.
//
// This file is free software; you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
// as published by the Free Software Foundation, in version 3 as it comes
// in the LICENSE.txt file of the Open vStorage OSE distribution.
//
// Open vStorage is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY of any kind.
/*global define */
define([
    'jquery', 'knockout',
    'ovs/generic', 'ovs/api'
], function($, ko, generic, api) {
    "use strict";
    return function(guid, parent) {
        var self = this;

        // Variables
        self.parent = parent;

        // Handles
        self.loadHandle = undefined;

        // Observables
        self.aliases      = ko.observable();
        self.filesystem   = ko.observable();
        self.guid         = ko.observable(guid);
        self.loaded       = ko.observable(false);
        self.loading      = ko.observable(false);
        self.mountpoint   = ko.observable();
        self.offset       = ko.observable().extend({ format: generic.formatBytes });
        self.relativeSize = ko.observable();
        self.roles        = ko.observableArray([]);
        self.size         = ko.observable().extend({ format: generic.formatBytes });
        self.state        = ko.observable();
        self.trigger      = ko.observable();
        self.usage        = ko.observable();

        // Functions
        self.fillData = function(data) {
            self.filesystem(data.filesystem);
            self.state(data.state);
            self.offset(data.offset);
            self.size(data.size);
            if (self.relativeSize() === undefined && parent.size !== undefined && parent.size.raw !== undefined && parent.size.raw() !== 0) {
                self.relativeSize(Math.round(data.size / parent.size.raw() * 100))
            }
            self.mountpoint(data.mountpoint);
            self.aliases(data.aliases);
            self.usage(generic.tryGet(data, 'usage', undefined));
            self.roles(data.roles);

            self.loaded(true);
            self.loading(false);
            self.trigger(generic.getTimestamp());
        };
        self.load = function() {
            return $.Deferred(function(deferred) {
                self.loading(true);
                if (generic.xhrCompleted(self.loadHandle)) {
                    self.loadHandle = api.get('diskpartitions/' + self.guid())
                        .done(function(data) {
                            self.fillData(data);
                            deferred.resolve();
                        })
                        .fail(deferred.reject)
                        .always(function() {
                            self.loading(false);
                        });
                } else {
                    deferred.reject();
                }
            }).promise();
        };
    };
});

File: ovs/dal/hybrids/diskpartition.py
# Copyright (C) 2016 iNuron NV
#
# This file is part of Open vStorage Open Source Edition (OSE),
# as available from
#
#      http://www.openvstorage.org and
#      http://www.openvstorage.com.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
# as published by the Free Software Foundation, in version 3 as it comes
# in the LICENSE.txt file of the Open vStorage OSE distribution.
#
# Open vStorage is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY of any kind.

"""
DiskPartition module
"""
from ovs.dal.dataobject import DataObject
from ovs.dal.structures import Property, Relation, Dynamic
from ovs.dal.hybrids.disk import Disk
from ovs.constants.roles import BACKEND, DB, DTL, SCRUB, WRITE


class DiskPartition(DataObject):
    """
    The DiskPartition class represents a partition on a physical Disk
    """
    ROLES = DataObject.enumerator('Role', [BACKEND, DB, DTL, SCRUB, WRITE])
    VIRTUAL_STORAGE_LOCATION = '/mnt/storage'

    __properties = [Property('filesystem', str, mandatory=False, doc='The filesystem used on the partition'),
                    Property('state', Disk.STATES.keys(), doc='State of the partition'),
                    Property('offset', int, doc='Offset of the partition'),
                    Property('size', int, doc='Size of the partition'),
                    Property('mountpoint', str, mandatory=False, doc='Mountpoint of the partition, None if not mounted'),
                    Property('aliases', list, doc='The partition aliases'),
                    Property('roles', list, default=[], doc='A list of claimed roles')]
    __relations = [Relation('disk', Disk, 'partitions')]
    __dynamics = [Dynamic('usage', list, 120),
                  Dynamic('folder', str, 3600)]

    def _usage(self):
        """
        A dict representing this partition's usage in a more user-friendly form
        """
        dataset = []
        for junction in self.storagedrivers:
            dataset.append({'type': 'storagedriver',
                            'role': junction.role,
                            'size': junction.size,
                            'relation': junction.storagedriver_guid,
                            'folder': junction.folder})
        return dataset

    def _folder(self):
        """
        Corrected mountpoint
        """
        return DiskPartition.VIRTUAL_STORAGE_LOCATION if self.mountpoint == '/' else self.mountpoint
